src/modules/submodules/consensus.py

```diff
@@ -0,0 +1,362 @@
+    def process_report(self, blockstamp: ReferenceBlockStamp) -> None:
+        """Builds and sends report for current frame with provided blockstamp."""
+        report_data = self.build_report(blockstamp)
+        logger.info({'msg': 'Build report.', 'value': report_data})
+
+        report_hash = self._get_report_hash(report_data)
+        logger.info({'msg': 'Calculate report hash.', 'value': report_hash})
+        self._process_report_hash(blockstamp, report_hash)
+        # Even if report hash transaction was failed we have to check if we can report data for current frame
+        self._process_report_data(blockstamp, report_data, report_hash)
+
+    def _process_report_hash(self, blockstamp: ReferenceBlockStamp, report_hash: HexBytes) -> None:
+        latest_blockstamp, member_info = self._get_latest_data()
+
+        # Check if current slot is higher than member slot + slots_delay
+        if not member_info.is_fast_lane:
+            if latest_blockstamp.slot_number < member_info.current_frame_ref_slot + member_info.fast_lane_length_slot:
+                logger.info({'msg': f'Member is not in fast lane, so report will be postponed for [{member_info.fast_lane_length_slot}] slots.'})
+                return None
+
+        if not member_info.is_report_member:
+            logger.info({'msg': 'Account can`t submit report hash.'})
+            return None
+
+        if HexBytes(member_info.current_frame_member_report) != report_hash:
+            logger.info({'msg': f'Send report hash. Consensus version: [{self.CONSENSUS_VERSION}]'})
+            self._send_report_hash(blockstamp, report_hash, self.CONSENSUS_VERSION)
+        else:
+            logger.info({'msg': 'Provided hash already submitted.'})
+
+        return None
+
+    def _process_report_data(self, blockstamp: ReferenceBlockStamp, report_data: tuple, report_hash: HexBytes):
+        latest_blockstamp, member_info = self._get_latest_data()
+
+        # If the quorum is ready to report data, the member should have opportunity to send report
+        if (
+            # If there is no quorum
+            member_info.current_frame_consensus_report == ZERO_HASH
+            # And member did not send the report
+            and HexBytes(member_info.current_frame_member_report) != report_hash
+        ):
+            logger.info({'msg': 'Quorum is not ready and member did not send the report hash.'})
+            return
+
+        if HexBytes(member_info.current_frame_consensus_report) != report_hash:
+            msg = 'Oracle`s hash differs from consensus report hash.'
+            logger.error({
+                'msg': msg,
+                'consensus_report_hash': str(HexBytes(member_info.current_frame_consensus_report)),
+                'report_hash': str(report_hash),
+            })
+            return
+
+        if self.is_main_data_submitted(latest_blockstamp):
+            logger.info({'msg': 'Main data already submitted.'})
+            return
```

Comment:
maybe remove this one?
we have this one below

Reply:
Done